### ISSUE TITLE
Use shared cache for stubs (?)

### ISSUE THREAD

**Issue by dantleech** [phpactor/phpactor#1816]
PHPStorm stubs are indexed in each project, it would be convenient to have them shared (and so the ability to merge indexes)

**Comment by mamazu** [comment]
I'd love to look into that. Is there any hints on where I could start. I am still unfamiliar with the structure of the project. So I just need a little info to get started.

**Comment by dantleech** [comment]
It's the src/Indexer package, but not sure if it will easily accommodate this feature, so might need some strategic refactoring to make it possible to implemt this.

**Comment by mamazu** [comment]
So from what I found out there is an `IndexJob` class that does the indexing. Where it has a filelist provider that it iterates over. Is there a reason there is only one IndexJob per project. Wouldn't it make sense to have multiple jobs running at the same time? Like one for the phpstubs and one for the project you are working on.

Secondly I couldn't find where the stubs are indexed. Is it something hardcoded or is it also configurable where they are?

**Comment by dantleech** [comment]
Possibly there could be multiple "index jobs"

Stubs are paths provided to the file list provider:

https://github.com/phpactor/phpactor/blob/2a13fb2cb50d59b167114af41f76b434633be717/lib/Indexer/Extension/IndexerExtension.php#L198-L211

**Comment by mamazu** [comment]
So my current plan would be to have the Stub Paths removed from the FileListProviders and have them be handled in a separate build job. But I am not sure that this is the right approach.

**Comment by dantleech** [comment]
You can try it, although we could keep the stub paths and have each as a separate job (potentially also separate progress trackers, but that would be later)

**Comment by devnix** [comment]
It would be cool to be able to generate a shared index for a project in a CI pipeline and then configure a download link to kickstart the indexing process 😄

**Comment by dantleech** [comment]
(fwiw what you suggest would already be possible though awkward :P )

**Comment by devnix** [comment]
[Is not a new idea](https://www.jetbrains.com/help/phpstorm/shared-indexes.html#install-shared-indexes-plugin), it can save a big time on big projects. I didn't even thought that indexes are already easily shareable, I might actually try something and share a gist 😄

**Comment by mamazu** [comment]
Yeah it definitely would also open up the oportunity to piggyback off of other projects if they have similar dependencies. I've been wanting to tackle this issue but I always got stuck on figuring out what classes do what in the indexing process.
